### PR TITLE
fix: add support for initializing global `array` variables

### LINKED ISSUE #3897: bug: external module's integer array element value is always 0
```fortran
module test
    implicit none
    integer(4), dimension(2) :: arr = [105, 109]
end module test

program a
    use test, only: arr
    implicit none
    integer :: i
    i = 1
    print *, arr(i)
end program a
```

the above program gives an output of '0' via LFortran, it should be '105'.

The issue is in `ASR->LLVM`, I found this issue while trying to fix https://github.com/lfortran/lfortran/issues/2969.

I've spent sometime trying to fix this issue, and would love to be the one to fix it.

### PR DESCRIPTION
**Pull request by kmr-srbh**
fixes #3897 

### The issue
The chief cause of the issue was `init_value` being a `nullptr` due to the use of `llvm::dyn_cast` to convert `tmp` of type `llvm::Value*` to `llvm::Constant*`. `tmp` is not a constant value here and so the cast returns a `nullptr` according to the definition of `llvm::dyn_cast<T>(U)`. `init_value` being a `nullptr` caused the initialization of the variable using a `zeroinitializer`, hence leading to the issue.

```llvm
@arr = global [4 x i32] zeroinitializer
```

### The fix
To fix the issue we initialize the global `array` variable manually using the `m_value` of the variable.
```fortran
module test
   implicit none
   integer(4), dimension(2) :: arr = [105, 109]
end module test

program a
   use test, only: arr
   implicit none
   integer :: i
   i = 1
   print *, arr(i)
end program a
```
```console
(base) saurabh-kumar@Awadh:~/Projects/System/lfortran$ lfortran ./examples/example.f90
105
```
### LLVM IR
For the provided MRE, here is the generated LLVM IR:
```llvm
; ModuleID = 'LFortran'
source_filename = "LFortran"

@arr = global [2 x i32] [i32 105, i32 109]
@0 = private unnamed_addr constant [2 x i8] c" \00", align 1
@1 = private unnamed_addr constant [2 x i8] c"\0A\00", align 1
@2 = private unnamed_addr constant [5 x i8] c"%d%s\00", align 1

define i32 @main(i32 %0, i8** %1) {
.entry:
  call void @_lpython_call_initial_functions(i32 %0, i8** %1)
  %i = alloca i32, align 4
  store i32 1, i32* %i, align 4
  %2 = load i32, i32* %i, align 4
  %3 = sub i32 %2, 1
  %4 = mul i32 1, %3
  %5 = add i32 0, %4
  %6 = getelementptr [2 x i32], [2 x i32]* @arr, i32 0, i32 %5
  %7 = load i32, i32* %6, align 4
  call void (i8*, ...) @_lfortran_printf(i8* getelementptr inbounds ([5 x i8], [5 x i8]* @2, i32 0, i32 0), i32 %7, i8* getelementptr inbounds ([2 x i8], [2 x i8]* @1, i32 0, i32 0))
  ret i32 0
}

declare void @_lpython_call_initial_functions(i32, i8**)

declare void @_lfortran_printf(i8*, ...)
```

### CODE REVIEW

**Comment by certik** [comment]
Go ahead and add an integration test that fails in master, but works in this PR.

**Comment by kmr-srbh** [comment]
Executing the above code currently throws the following error. 

```console
(lf) saurabh-kumar@Awadh:~/Projects/System/lfortran$ lfortran ./examples/example.f90
LLVM ERROR: invalid number of bytes
Traceback (most recent call last):
  Binary file "/home/saurabh-kumar/Projects/System/lfortran/src/bin/lfortran", in _start()
  File "./csu/../csu/libc-start.c", line 360, in __libc_start_main_impl()
  File "./csu/../sysdeps/x86/libc-start.c", line 58, in __libc_start_call_main()
  File "/home/saurabh-kumar/Projects/System/lfortran/src/bin/lfortran.cpp", line 2465, in main()
    return main_app(argc, argv);
  File "/home/saurabh-kumar/Projects/System/lfortran/src/bin/lfortran.cpp", line 2426, in main_app(int, char**)
    err = compile_to_object_file(arg_file, tmp_o, false,
  File "/home/saurabh-kumar/Projects/System/lfortran/src/bin/lfortran.cpp", line 965, in compile_to_object_file()
    e.save_object_file(*(m->m_m), outfile);
  File "/home/saurabh-kumar/Projects/System/lfortran/src/libasr/codegen/evaluator.cpp", line 359, in LCompilers::LLVMEvaluator::save_object_file(llvm::Module&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    pass.run(m);
  Binary file "/home/saurabh-kumar/Projects/System/lfortran/src/bin/lfortran", in llvm::legacy::PassManagerImpl::run(llvm::Module&)
  File "LegacyPassManager.cpp", line 0, in llvm::FPPassManager::doFinalization(llvm::Module&) [clone .localalias]
  Binary file "/home/saurabh-kumar/Projects/System/lfortran/src/bin/lfortran", in llvm::AsmPrinter::doFinalization(llvm::Module&)
  Binary file "/home/saurabh-kumar/Projects/System/lfortran/src/bin/lfortran", in llvm::MCStreamer::Finish()
  Binary file "/home/saurabh-kumar/Projects/System/lfortran/src/bin/lfortran", in llvm::MCAssembler::Finish()
  Binary file "/home/saurabh-kumar/Projects/System/lfortran/src/bin/lfortran", in llvm::MCAssembler::layout(llvm::MCAsmLayout&)
  Binary file "/home/saurabh-kumar/Projects/System/lfortran/src/bin/lfortran", in llvm::MCAssembler::finishLayout(llvm::MCAsmLayout&)
  Binary file "/home/saurabh-kumar/Projects/System/lfortran/src/bin/lfortran", in llvm::MCAssembler::computeFragmentSize(llvm::MCAsmLayout const&, llvm::MCFragment const&) const
  Binary file "/home/saurabh-kumar/Projects/System/lfortran/src/bin/lfortran", in llvm::MCContext::reportError(llvm::SMLoc, llvm::Twine const&)
  Binary file "/home/saurabh-kumar/Projects/System/lfortran/src/bin/lfortran", in llvm::report_fatal_error(llvm::Twine const&, bool)
  File "./stdlib/./stdlib/abort.c", line 79, in __GI_abort()
  File "./signal/../sysdeps/posix/raise.c", line 26, in __GI_raise()
  File "./nptl/./nptl/pthread_kill.c", line 43, in __pthread_kill_implementation()
  Binary file "/lib/x86_64-linux-gnu/libc.so.6", in __GI___sigaction()
Abort: Signal SIGABRT (abort) received

Aborted (core dumped)
```

I feel that the generated LLVM IR is correct and this error is caused due to incorrect memory alignment. I tried very hard to find the place of the issue, but could not. As this is a new implementation with global array declaration, I am guessing there might be something more required. Could somebody please guide me on this?

**Comment by certik** [comment]
Note sure if the LLVM code is correct. You can try to load it into `llvm-opt` directly (save the LLVM code into `a.ll` and do `llvm-opt a.ll`) and try to compile it by hand. See if you get any errors.

**Comment by kmr-srbh** [comment]
With the latest commit, we now successfully handle the initialization of global arrays.
```fortran
module test
   implicit none
   integer(4), dimension(2) :: arr = [105, 109]
end module test

program a
   use test, only: arr
   implicit none
   integer :: i
   i = 1
   print *, arr(i)
end program a
```
```console
(base) saurabh-kumar@Awadh:~/Projects/System/lfortran$ lfortran ./examples/example.f90
105
```

**Comment by Thirumalai-Shaktivel** [comment]
Does the following work?
```fortran
module test
   implicit none
   integer(4), dimension(2) :: arr = [105, 109]
end module test

program a
   use test, only: arr
   implicit none
   integer :: i
   i = 1
   print *, arr
end program a
```

**Comment by Thirumalai-Shaktivel** [comment]
Or 
```fortran
program test
integer :: x(3), y(3)
x = 3
y = x + x
print *, y
end program
```

**Comment by kmr-srbh** [comment]
> Does the following work?

Both the above code samples work. I have similar test cases tested too.

```console
(base) saurabh-kumar@Awadh:~/Projects/System/lfortran$ lfortran ./examples/example.f90
105 109 
(base) saurabh-kumar@Awadh:~/Projects/System/lfortran$ lfortran ./examples/example.f90
108 112 
105 109 
```

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 1]
```diff
@@ -2645,14 +2645,43 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
...
+                    for (size_t i = 0; i < arr_const_size; i++) {
+                        ASR::expr_t* elem = ASRUtils::fetch_ArrayConstant_value(al, arr_const, i);
+                        a_kind = ASRUtils::extract_kind_from_ttype_t(ASRUtils::expr_type(elem));
+                        if (ASR::is_a<ASR::IntegerConstant_t>(*elem)) {
+                            ASR::IntegerConstant_t* int_const = ASR::down_cast<ASR::IntegerConstant_t>(elem);
+                            arr_elements.push_back(llvm::ConstantInt::get(
+                                context, llvm::APInt(8 * a_kind, int_const->m_n)));
+                        } else if (ASR::is_a<ASR::RealConstant_t>(*elem)) {
+                            ASR::RealConstant_t* real_const = ASR::down_cast<ASR::RealConstant_t>(elem);
+                            if (8 * a_kind == 32) {
```
Comment (certik) [inline src/libasr/codegen/asr_to_llvm.cpp:2667]: This is more readable, we use 4 for single and 8 for double:
```suggestion
                            if (a_kind == 4) {
```

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 2]
```diff
@@ -2645,14 +2645,43 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
...
+                                arr_elements.push_back(llvm::ConstantFP::get(
+                                    context, llvm::APFloat((float) real_const->m_r)));
+                            } else if (8 * a_kind == 64) {
```
Comment (certik) [inline src/libasr/codegen/asr_to_llvm.cpp:2670]:
```suggestion
                            } else if (a_kind == 8) {
```

**Review by certik — APPROVED** [review]
I think this looks good to me, thanks!

**Pull request merged**